File: 39.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        Solution.backtrack(result, [], 0, candidates, 0, target)
        return result

    @staticmethod
    def backtrack(result, tmp, start, nums, s, target):
        if s > target:
            return
        elif s == target:
            result.append(tmp[::])
        else:
            for index in range(start, len(nums)):
                num = nums[index]
                tmp.append(num)
                Solution.backtrack(result, tmp, index, nums, s + num, target)
                tmp.pop()


if __name__ == '__main__':
    print(Solution().combinationSum([2, 3, 6, 7], 7))
    print(Solution().combinationSum([2, 3, 5], 8))

File: 39.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        combinationSum(candidates, target, result, new int[candidates.length], 0);
        return result;
    }

    private void combinationSum(int[] candidates, int target, List<List<Integer>> result, int[] tmp, int startIndex) {
        if (target == 0) {
            List<Integer> subResult = new ArrayList<>();
            for (int i = 0; i < tmp.length; i++) {
                for (int j = 0; j < tmp[i]; j++) {
                    subResult.add(candidates[i]);
                }
            }
            result.add(subResult);
            return;
        }
        if (startIndex >= candidates.length) {
            return;
        }
        int num = 0;
        while (num <= target) {
            combinationSum(candidates, target - num, result, tmp, startIndex + 1);
            num += candidates[startIndex];
            tmp[startIndex]++;
        }
        tmp[startIndex] = 0;
    }
}
